src/main/scala/akka/persistence/jdbc/query/JournalSequenceActor.scala

```diff
@@ -57,47 +57,34 @@ class JournalSequenceActor(readJournalDao: ReadJournalDao, config: JournalSequen
+
     case ScheduleAssumeMaxOrderingId(max) =>
       // All elements smaller than max can be assumed missing after this delay
       val delay = queryDelay * maxTries
       timers.startSingleTimer(key = AssumeMaxOrderingIdTimerKey, AssumeMaxOrderingId(max), delay)
+
     case AssumeMaxOrderingId(max) =>
       if (currentMaxOrdering < max) {
         context.become(receive(max, missingByCounter, moduloCounter, previousDelay))
       }
+
     case GetMaxOrderingId =>
       sender() ! MaxOrderingId(currentMaxOrdering)
+
     case QueryOrderingIds =>
-      readJournalDao.journalSequence(currentMaxOrdering, batchSize).runWith(Sink.seq)
-        .map(result => NewOrderingIds(result)) pipeTo self
-    case NewOrderingIds(elements) =>
-      val givenUp = missingByCounter.getOrElse(moduloCounter, Set.empty)
-      val (nextmax, _, missingElems) = elements.foldLeft[(OrderingId, OrderingId, Set[OrderingId])](currentMaxOrdering, currentMaxOrdering, Set.empty) {
-        case ((currentMax, previousElement, missing), elem) =>
-          val newMissing = if (previousElement + 1 < elem && !givenUp(elem)) {
-            val currentlyMissing = previousElement + 1 until elem
-            def alreadyMissing(e: Long) = missingByCounter.values.exists(_.contains(e))
-            missing ++ currentlyMissing.filterNot(alreadyMissing)
-          } else missing
-          val newMax =
-            if (currentMax + 1 == elem) elem
-            else if ((currentMax + 1).until(elem).forall(givenUp.contains)) elem
-            else currentMax
-          (newMax, elem, newMissing)
-      }
-      val newMissingByCounter = (missingByCounter + (moduloCounter -> missingElems)).map {
-        case (key, value) =>
-          key -> value.filter(missingId => missingId > nextmax)
-      }
-      if (nextmax - currentMaxOrdering >= batchSize && newMissingByCounter.values.forall(_.isEmpty)) {
-        // Many elements have been retrieved but none are missing
-        // We can query again immediately, as this allows the actor to rapidly retrieve the real max ordering
-        self ! QueryOrderingIds
-        context.become(receive(nextmax, newMissingByCounter, moduloCounter))
-      } else {
-        scheduleQuery(queryDelay)
-        context.become(receive(nextmax, newMissingByCounter, (moduloCounter + 1) % maxTries))
-      }
+      readJournalDao
+        .journalSequence(currentMaxOrdering, batchSize)
+        .runWith(Sink.seq)
+        .map(result => NewOrderingIds(currentMaxOrdering, result)) pipeTo self
+
+    case NewOrderingIds(originalOffset, _) if originalOffset < currentMaxOrdering =>
+      // search was done using an offset that became obsolete in the meantime
+      // therefore we start a new query
+      self ! QueryOrderingIds
+
+    case NewOrderingIds(_, elements) =>
+      findGaps(elements, currentMaxOrdering, missingByCounter, moduloCounter)
```

Comment:
I found better to move the main algorithm code to it's one method. That gives better visibility when we want to have a glimpse on all the messages that the actor can handle.

Reply:
Looks good.

Perhaps it would be a good idea to move the `context.become(receive(...))` out of the findGaps method and into this receive function again. That makes is more clear that we are changing the state of the actor.

Reply:
I agree, that would be much clear if all `context.become` appear on the same level.

Reply:
I just tried and unfortunately it depends on 3 values that are only in the scope of `findGaps`.
* `noGapsFound && isFullBatch` - Boolean (two values but reducible to 1)
* `nextMax` - Long
* `newMissingByCounter` - Map[Int, Set[JournalSequenceActor.OrderingId]]

Not sure if returning a case class or a tuple3 (Boolean, Long, Map[Int, Set[JournalSequenceActor.OrderingId]]) will add clarity to it.